homeassistant/components/tplink/vacuum.py

```diff
@@ -0,0 +1,158 @@
+"""Support for TPLink vacuum."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, cast
+
+from kasa import Device, Feature, Module
+from kasa.smart.modules.clean import Clean, Status
+from kasa.smart.modules.speaker import Speaker
+
+from homeassistant.components.vacuum import (
+    DOMAIN as VACUUM_DOMAIN,
+    StateVacuumEntity,
+    StateVacuumEntityDescription,
+    VacuumActivity,
+    VacuumEntityFeature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TPLinkConfigEntry
+from .coordinator import TPLinkDataUpdateCoordinator
+from .entity import (
+    CoordinatedTPLinkModuleEntity,
+    TPLinkModuleEntityDescription,
+    async_refresh_after,
+)
+
+# Coordinator is used to centralize the data updates
+# For actions the integration handles locking of concurrent device request
+PARALLEL_UPDATES = 0
+
+# Upstream state to VacuumActivity
+STATUS_TO_ACTIVITY = {
+    Status.Idle: VacuumActivity.IDLE,
+    Status.Cleaning: VacuumActivity.CLEANING,
+    Status.GoingHome: VacuumActivity.RETURNING,
+    Status.Charging: VacuumActivity.DOCKED,
+    Status.Charged: VacuumActivity.DOCKED,
+    Status.Undocked: VacuumActivity.IDLE,
+    Status.Paused: VacuumActivity.PAUSED,
+    Status.Error: VacuumActivity.ERROR,
+}
+
+
+@dataclass(frozen=True, kw_only=True)
+class TPLinkVacuumEntityDescription(
+    StateVacuumEntityDescription, TPLinkModuleEntityDescription
+):
+    """Base class for vacuum entity description."""
+
+
+VACUUM_DESCRIPTIONS: tuple[TPLinkVacuumEntityDescription, ...] = (
+    TPLinkVacuumEntityDescription(
+        key="vacuum",
+        exists_fn=lambda dev, _: Module.Clean in dev.modules,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up vacuum entities."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    device = parent_coordinator.device
+
+    known_child_device_ids: set[str] = set()
+    first_check = True
```

Comment:
Something for the future: it would be nicer to make this a bit more DRY since `first_check` is in almost every platform